### PR TITLE
update: new data source aws_msk_bootstrap_brokers

### LINKED ISSUE #28005: [Enhancement]: aws_msk_serverless_cluster to return bootstrap server url
### Description

The aws_msk_serverless_cluster recently added doesn't return the URL of the created cluster.

Now this ticket will need to be on hold, pending on aws to provide a way to programmatically retrieve that information.

As far as I can see it's not available: [describe-cluster-v2](https://docs.aws.amazon.com/cli/latest/reference/kafka/describe-cluster-v2.html) returned data doesn't have the information.

And there are no method related to client-information:
https://docs.aws.amazon.com/cli/latest/reference/kafka/index.html

### Affected Resource(s) and/or Data Source(s)

* aws_msk_serverless_cluster

### Potential Terraform Configuration

```terraform
* aws_msk_serverless_cluster.client_information
```

### References

https://github.com/hashicorp/terraform-provider-aws/issues/22058

### Would you like to implement a fix?

No

### LINKED ISSUE #32472: [New Data Source]: MSK Bootstrap Brokers
### Description

We need to be able to access the `BootstrapBrokerStringSaslIam` from a provisioned `aws_msk_serverless_cluster`. 

The current datasources (https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/msk_broker_nodes) do not work with the serverless cluster. 

> aws kafka list-nodes --cluster-arn "<redacted>" --region "<redacted>" 
> 
> An error occurred (BadRequestException) when calling the ListNodes operation: This operation cannot be performed on serverless clusters.

There is what i believe is a new command `get-bootstrap-brokers` where the values can be obtained via the following command https://awscli.amazonaws.com/v2/documentation/api/latest/reference/kafka/get-bootstrap-brokers.html

```
aws kafka get-bootstrap-brokers --cluster-arn <ARN> --region <REGION
```

### Requested Resource(s) and/or Data Source(s)

`aws_msk_boostrap_brokers`

### Potential Terraform Configuration

```terraform
data "aws_msk_bootstrap_brokers" "example" {
  cluster_arn = aws_msk_cluster.example.arn
}
```

### References

https://docs.aws.amazon.com/sdk-for-go/api/service/kafka/#Kafka.GetBootstrapBrokers

### Would you like to implement a fix?

Yes

### PR DESCRIPTION
**Pull request by kaykhan**
### Description

Attempting to add a datasource to fetch the boostrap broker url `BootstrapBrokerStringPublicSaslIam` for a serverless msk cluster.

Existing datasources do not work with msk serverless cluster. 

New command `GetBootstrapBrokers` works for both serverless and original msk cluster

https://docs.aws.amazon.com/sdk-for-go/api/service/kafka/#Kafka.GetBootstrapBrokers
https://docs.aws.amazon.com/cli/latest/reference/kafka/get-bootstrap-brokers.html

### Relations

Closes #32472
Closes #28005

### References

https://docs.aws.amazon.com/sdk-for-go/api/service/kafka/#Kafka.GetBootstrapBrokers

```
aws kafka get-bootstrap-brokers --cluster-arn <ARN> --region <REGION>
```

### Output from Acceptance Testing

```
$ make testacc TESTS=TestAccXXX PKG=ec2

...
```

### CODE REVIEW

**Comment by joaocc** [comment]
Hi
Any updates on merging this branch?
We would love to have this deployed, as it seems we don't have any other way to use terraform to connect to a serverless cluster.
Thanks

**Comment by apd-bscally** [comment]
Hi, echoing the previous comment, looking for this functionality

**Comment by kaykhan** [comment]
[ewbankkit](https://github.com/ewbankkit) [justinretzolk](https://github.com/justinretzolk) any chance i could get help with this one

**Comment by NexRX** [comment]
Hi all, is this still being developed? Would be awesome to get this and its great to see the effort.  🤩

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% ACCTEST_TIMEOUT=720m make testacc TESTARGS='-run=TestAccKafkaBootstrapBrokersDataSource_\|TestAccKafkaBrokerNodesDataSource_\|TestAccKafkaClusterDataSource_\|TestAccKafkaClusterPolicy_\|TestAccKafkaConfigurationDataSource_\|TestAccKafkaKafkaVersionDataSource_\|TestAccKafkaVPCConnectionDataSource_' PKG=kafka ACCTEST_PARALLELISM=3
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/kafka/... -v -count 1 -parallel 3  -run=TestAccKafkaBootstrapBrokersDataSource_\|TestAccKafkaBrokerNodesDataSource_\|TestAccKafkaClusterDataSource_\|TestAccKafkaClusterPolicy_\|TestAccKafkaConfigurationDataSource_\|TestAccKafkaKafkaVersionDataSource_\|TestAccKafkaVPCConnectionDataSource_ -timeout 720m
=== RUN   TestAccKafkaBootstrapBrokersDataSource_basic
=== PAUSE TestAccKafkaBootstrapBrokersDataSource_basic
=== RUN   TestAccKafkaBrokerNodesDataSource_basic
=== PAUSE TestAccKafkaBrokerNodesDataSource_basic
=== RUN   TestAccKafkaClusterDataSource_basic
=== PAUSE TestAccKafkaClusterDataSource_basic
=== RUN   TestAccKafkaClusterPolicy_basic
=== PAUSE TestAccKafkaClusterPolicy_basic
=== RUN   TestAccKafkaClusterPolicy_disappears
=== PAUSE TestAccKafkaClusterPolicy_disappears
=== RUN   TestAccKafkaClusterPolicy_update
=== PAUSE TestAccKafkaClusterPolicy_update
=== RUN   TestAccKafkaConfigurationDataSource_name
=== PAUSE TestAccKafkaConfigurationDataSource_name
=== RUN   TestAccKafkaKafkaVersionDataSource_basic
=== PAUSE TestAccKafkaKafkaVersionDataSource_basic
=== RUN   TestAccKafkaKafkaVersionDataSource_preferred
=== PAUSE TestAccKafkaKafkaVersionDataSource_preferred
=== RUN   TestAccKafkaVPCConnectionDataSource_basic
=== PAUSE TestAccKafkaVPCConnectionDataSource_basic
=== CONT  TestAccKafkaBootstrapBrokersDataSource_basic
=== CONT  TestAccKafkaVPCConnectionDataSource_basic
=== CONT  TestAccKafkaKafkaVersionDataSource_preferred
--- PASS: TestAccKafkaKafkaVersionDataSource_preferred (13.84s)
=== CONT  TestAccKafkaClusterPolicy_basic
--- PASS: TestAccKafkaBootstrapBrokersDataSource_basic (1995.28s)
=== CONT  TestAccKafkaKafkaVersionDataSource_basic
--- PASS: TestAccKafkaKafkaVersionDataSource_basic (11.02s)
=== CONT  TestAccKafkaConfigurationDataSource_name
--- PASS: TestAccKafkaConfigurationDataSource_name (45.21s)
=== CONT  TestAccKafkaClusterPolicy_update
--- PASS: TestAccKafkaClusterPolicy_basic (6209.08s)
=== CONT  TestAccKafkaClusterPolicy_disappears
--- PASS: TestAccKafkaVPCConnectionDataSource_basic (6225.98s)
=== CONT  TestAccKafkaBrokerNodesDataSource_basic
--- PASS: TestAccKafkaBrokerNodesDataSource_basic (2253.03s)
=== CONT  TestAccKafkaClusterDataSource_basic
--- PASS: TestAccKafkaClusterPolicy_update (6474.94s)
--- PASS: TestAccKafkaClusterDataSource_basic (1987.09s)
--- PASS: TestAccKafkaClusterPolicy_disappears (6714.33s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/kafka	12938.119s
```
```console
% ACCTEST_TIMEOUT=720m make testacc TESTARGS='-run=TestAccKafkaServerlessCluster_\|TestAccKafkaVPCConnection_\|TestAccKafkaScramSecretAssociation_\|TestAccKafkaConfiguration_' PKG=kafka ACCTEST_PARALLELISM=3
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/kafka/... -v -count 1 -parallel 3  -run=TestAccKafkaServerlessCluster_\|TestAccKafkaVPCConnection_\|TestAccKafkaScramSecretAssociation_\|TestAccKafkaConfiguration_ -timeout 720m
=== RUN   TestAccKafkaConfiguration_basic
=== PAUSE TestAccKafkaConfiguration_basic
=== RUN   TestAccKafkaConfiguration_disappears
=== PAUSE TestAccKafkaConfiguration_disappears
=== RUN   TestAccKafkaConfiguration_description
=== PAUSE TestAccKafkaConfiguration_description
=== RUN   TestAccKafkaConfiguration_kafkaVersions
=== PAUSE TestAccKafkaConfiguration_kafkaVersions
=== RUN   TestAccKafkaConfiguration_serverProperties
=== PAUSE TestAccKafkaConfiguration_serverProperties
=== RUN   TestAccKafkaScramSecretAssociation_basic
=== PAUSE TestAccKafkaScramSecretAssociation_basic
=== RUN   TestAccKafkaScramSecretAssociation_update
=== PAUSE TestAccKafkaScramSecretAssociation_update
=== RUN   TestAccKafkaScramSecretAssociation_disappears
=== PAUSE TestAccKafkaScramSecretAssociation_disappears
=== RUN   TestAccKafkaScramSecretAssociation_Disappears_cluster
=== PAUSE TestAccKafkaScramSecretAssociation_Disappears_cluster
=== RUN   TestAccKafkaServerlessCluster_basic
=== PAUSE TestAccKafkaServerlessCluster_basic
=== RUN   TestAccKafkaServerlessCluster_disappears
=== PAUSE TestAccKafkaServerlessCluster_disappears
=== RUN   TestAccKafkaServerlessCluster_tags
=== PAUSE TestAccKafkaServerlessCluster_tags
=== RUN   TestAccKafkaServerlessCluster_securityGroup
=== PAUSE TestAccKafkaServerlessCluster_securityGroup
=== RUN   TestAccKafkaVPCConnection_basic
=== PAUSE TestAccKafkaVPCConnection_basic
=== RUN   TestAccKafkaVPCConnection_tags
=== PAUSE TestAccKafkaVPCConnection_tags
=== RUN   TestAccKafkaVPCConnection_disappears
=== PAUSE TestAccKafkaVPCConnection_disappears
=== CONT  TestAccKafkaConfiguration_basic
=== CONT  TestAccKafkaScramSecretAssociation_Disappears_cluster
=== CONT  TestAccKafkaConfiguration_serverProperties
--- PASS: TestAccKafkaConfiguration_basic (23.35s)
=== CONT  TestAccKafkaServerlessCluster_securityGroup
--- PASS: TestAccKafkaConfiguration_serverProperties (38.68s)
=== CONT  TestAccKafkaVPCConnection_disappears
--- PASS: TestAccKafkaServerlessCluster_securityGroup (181.34s)
=== CONT  TestAccKafkaVPCConnection_tags
--- PASS: TestAccKafkaScramSecretAssociation_Disappears_cluster (2044.31s)
=== CONT  TestAccKafkaVPCConnection_basic
--- PASS: TestAccKafkaVPCConnection_tags (6153.11s)
=== CONT  TestAccKafkaScramSecretAssociation_update
--- PASS: TestAccKafkaVPCConnection_disappears (6326.22s)
=== CONT  TestAccKafkaScramSecretAssociation_disappears
--- PASS: TestAccKafkaScramSecretAssociation_disappears (1784.09s)
=== CONT  TestAccKafkaConfiguration_description
--- PASS: TestAccKafkaConfiguration_description (79.34s)
=== CONT  TestAccKafkaConfiguration_kafkaVersions
--- PASS: TestAccKafkaConfiguration_kafkaVersions (22.31s)
=== CONT  TestAccKafkaScramSecretAssociation_basic
--- PASS: TestAccKafkaScramSecretAssociation_update (2163.80s)
=== CONT  TestAccKafkaConfiguration_disappears
--- PASS: TestAccKafkaConfiguration_disappears (22.86s)
=== CONT  TestAccKafkaServerlessCluster_disappears
--- PASS: TestAccKafkaVPCConnection_basic (6613.02s)
=== CONT  TestAccKafkaServerlessCluster_tags
--- PASS: TestAccKafkaServerlessCluster_disappears (187.21s)
=== CONT  TestAccKafkaServerlessCluster_basic
--- PASS: TestAccKafkaServerlessCluster_tags (218.01s)
--- PASS: TestAccKafkaServerlessCluster_basic (260.81s)
--- PASS: TestAccKafkaScramSecretAssociation_basic (2066.57s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/kafka	10419.727s
```
```console
% ACCTEST_TIMEOUT=720m make testacc TESTARGS='-run=TestAccKafkaCluster_' PKG=kafka ACCTEST_PARALLELISM=3
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/kafka/... -v -count 1 -parallel 3  -run=TestAccKafkaCluster_ -timeout 720m
=== RUN   TestAccKafkaCluster_basic
=== PAUSE TestAccKafkaCluster_basic
=== RUN   TestAccKafkaCluster_disappears
=== PAUSE TestAccKafkaCluster_disappears
=== RUN   TestAccKafkaCluster_tags
=== PAUSE TestAccKafkaCluster_tags
=== RUN   TestAccKafkaCluster_BrokerNodeGroupInfo_storageInfo
=== PAUSE TestAccKafkaCluster_BrokerNodeGroupInfo_storageInfo
=== RUN   TestAccKafkaCluster_BrokerNodeGroupInfo_instanceType
=== PAUSE TestAccKafkaCluster_BrokerNodeGroupInfo_instanceType
=== RUN   TestAccKafkaCluster_BrokerNodeGroupInfo_publicAccessSASLIAM
=== PAUSE TestAccKafkaCluster_BrokerNodeGroupInfo_publicAccessSASLIAM
=== RUN   TestAccKafkaCluster_BrokerNodeGroupInfo_vpcConnectivity
=== PAUSE TestAccKafkaCluster_BrokerNodeGroupInfo_vpcConnectivity
=== RUN   TestAccKafkaCluster_ClientAuthenticationSASL_scram
=== PAUSE TestAccKafkaCluster_ClientAuthenticationSASL_scram
=== RUN   TestAccKafkaCluster_ClientAuthenticationSASL_iam
=== PAUSE TestAccKafkaCluster_ClientAuthenticationSASL_iam
=== RUN   TestAccKafkaCluster_ClientAuthenticationTLS_certificateAuthorityARNs
=== PAUSE TestAccKafkaCluster_ClientAuthenticationTLS_certificateAuthorityARNs
=== RUN   TestAccKafkaCluster_ClientAuthenticationTLS_initiallyNoAuthentication
=== PAUSE TestAccKafkaCluster_ClientAuthenticationTLS_initiallyNoAuthentication
=== RUN   TestAccKafkaCluster_Info_revision
=== PAUSE TestAccKafkaCluster_Info_revision
=== RUN   TestAccKafkaCluster_EncryptionInfo_encryptionAtRestKMSKeyARN
=== PAUSE TestAccKafkaCluster_EncryptionInfo_encryptionAtRestKMSKeyARN
=== RUN   TestAccKafkaCluster_EncryptionInfoEncryptionInTransit_clientBroker
=== PAUSE TestAccKafkaCluster_EncryptionInfoEncryptionInTransit_clientBroker
=== RUN   TestAccKafkaCluster_EncryptionInfoEncryptionInTransit_inCluster
=== PAUSE TestAccKafkaCluster_EncryptionInfoEncryptionInTransit_inCluster
=== RUN   TestAccKafkaCluster_enhancedMonitoring
=== PAUSE TestAccKafkaCluster_enhancedMonitoring
=== RUN   TestAccKafkaCluster_numberOfBrokerNodes
=== PAUSE TestAccKafkaCluster_numberOfBrokerNodes
=== RUN   TestAccKafkaCluster_openMonitoring
=== PAUSE TestAccKafkaCluster_openMonitoring
=== RUN   TestAccKafkaCluster_storageMode
=== PAUSE TestAccKafkaCluster_storageMode
=== RUN   TestAccKafkaCluster_loggingInfo
=== PAUSE TestAccKafkaCluster_loggingInfo
=== RUN   TestAccKafkaCluster_kafkaVersionUpgrade
=== PAUSE TestAccKafkaCluster_kafkaVersionUpgrade
=== RUN   TestAccKafkaCluster_kafkaVersionDowngrade
=== PAUSE TestAccKafkaCluster_kafkaVersionDowngrade
=== RUN   TestAccKafkaCluster_kafkaVersionUpgradeWithInfo
=== PAUSE TestAccKafkaCluster_kafkaVersionUpgradeWithInfo
=== CONT  TestAccKafkaCluster_basic
=== CONT  TestAccKafkaCluster_EncryptionInfo_encryptionAtRestKMSKeyARN
=== CONT  TestAccKafkaCluster_storageMode
--- PASS: TestAccKafkaCluster_EncryptionInfo_encryptionAtRestKMSKeyARN (2072.83s)
=== CONT  TestAccKafkaCluster_kafkaVersionUpgradeWithInfo
--- PASS: TestAccKafkaCluster_basic (2082.98s)
=== CONT  TestAccKafkaCluster_kafkaVersionDowngrade
--- PASS: TestAccKafkaCluster_storageMode (3072.02s)
=== CONT  TestAccKafkaCluster_kafkaVersionUpgrade
--- PASS: TestAccKafkaCluster_kafkaVersionDowngrade (4551.30s)
=== CONT  TestAccKafkaCluster_loggingInfo
--- PASS: TestAccKafkaCluster_kafkaVersionUpgradeWithInfo (6826.36s)
=== CONT  TestAccKafkaCluster_ClientAuthenticationTLS_certificateAuthorityARNs
--- PASS: TestAccKafkaCluster_loggingInfo (2428.23s)
=== CONT  TestAccKafkaCluster_Info_revision
--- PASS: TestAccKafkaCluster_kafkaVersionUpgrade (7003.90s)
=== CONT  TestAccKafkaCluster_ClientAuthenticationTLS_initiallyNoAuthentication
--- PASS: TestAccKafkaCluster_ClientAuthenticationTLS_certificateAuthorityARNs (2293.01s)
=== CONT  TestAccKafkaCluster_ClientAuthenticationSASL_iam
--- PASS: TestAccKafkaCluster_Info_revision (2690.34s)
=== CONT  TestAccKafkaCluster_ClientAuthenticationSASL_scram
--- PASS: TestAccKafkaCluster_ClientAuthenticationTLS_initiallyNoAuthentication (4381.02s)
=== CONT  TestAccKafkaCluster_BrokerNodeGroupInfo_storageInfo
--- PASS: TestAccKafkaCluster_ClientAuthenticationSASL_iam (4402.25s)
=== CONT  TestAccKafkaCluster_BrokerNodeGroupInfo_publicAccessSASLIAM
--- PASS: TestAccKafkaCluster_ClientAuthenticationSASL_scram (4011.36s)
=== CONT  TestAccKafkaCluster_BrokerNodeGroupInfo_instanceType
--- PASS: TestAccKafkaCluster_BrokerNodeGroupInfo_storageInfo (2466.10s)
=== CONT  TestAccKafkaCluster_enhancedMonitoring
--- PASS: TestAccKafkaCluster_enhancedMonitoring (2254.81s)
=== CONT  TestAccKafkaCluster_openMonitoring
--- PASS: TestAccKafkaCluster_openMonitoring (1200.96s)
=== CONT  TestAccKafkaCluster_numberOfBrokerNodes
--- PASS: TestAccKafkaCluster_BrokerNodeGroupInfo_publicAccessSASLIAM (4821.43s)
=== CONT  TestAccKafkaCluster_tags
--- PASS: TestAccKafkaCluster_BrokerNodeGroupInfo_instanceType (5988.85s)
=== CONT  TestAccKafkaCluster_EncryptionInfoEncryptionInTransit_inCluster
--- PASS: TestAccKafkaCluster_tags (1962.64s)
=== CONT  TestAccKafkaCluster_disappears
--- PASS: TestAccKafkaCluster_numberOfBrokerNodes (3502.55s)
=== CONT  TestAccKafkaCluster_BrokerNodeGroupInfo_vpcConnectivity
--- PASS: TestAccKafkaCluster_EncryptionInfoEncryptionInTransit_inCluster (2134.46s)
=== CONT  TestAccKafkaCluster_EncryptionInfoEncryptionInTransit_clientBroker
--- PASS: TestAccKafkaCluster_disappears (1546.30s)
--- PASS: TestAccKafkaCluster_EncryptionInfoEncryptionInTransit_clientBroker (2045.99s)
--- PASS: TestAccKafkaCluster_BrokerNodeGroupInfo_vpcConnectivity (5949.54s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/kafka	27314.439s
```
```console
% ACCTEST_TIMEOUT=720m make testacc TESTARGS='-run=TestAccKafkaReplicator_update' PKG=kafka
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/kafka/... -v -count 1 -parallel 20  -run=TestAccKafkaReplicator_update -timeout 720m
=== RUN   TestAccKafkaReplicator_update
=== PAUSE TestAccKafkaReplicator_update
=== CONT  TestAccKafkaReplicator_update
--- PASS: TestAccKafkaReplicator_update (6791.22s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/kafka	6797.636s
```

**Comment by ewbankkit** [comment]
@kaykhan Thanks for the contribution :tada: :clap:.

**Comment by kaykhan** [comment]
> @kaykhan Thanks for the contribution 🎉 👏.

appreciate your help, thanks!

File: website/docs/d/msk_bootstrap_brokers.html.markdown [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+---
+
+# Data Source: aws_msk_bootstrap_brokers
+
+Get a list of brokers that a client application can use to bootstrap.
+
+## Example Usage
+
+```terraform
+data "aws_msk_cluster" "example" {
```
Comment (yermulnik) [inline website/docs/d/msk_bootstrap_brokers.html.markdown:16]: `aws_msk_cluster` should probably be `aws_msk_bootstrap_brokers`?

File: website/docs/d/msk_bootstrap_brokers.html.markdown [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+subcategory: "Managed Streaming for Kafka"
+layout: "aws"
+page_title: "AWS: aws_msk_bootstrap_brokers"
+description: |-
+    Get a list of brokers that a client application can use to bootstrap.
+---
+
+# Data Source: aws_msk_bootstrap_brokers
+
+Get a list of brokers that a client application can use to bootstrap.
```
Comment (yermulnik) [inline website/docs/d/msk_bootstrap_brokers.html.markdown:11]: May probably make sense to add a note on what's the difference as opposite to `aws_msk_broker_nodes` data source? 🤔

**Pull request merged**